### PR TITLE
Add graylog service 

### PR DESCRIPTION
**Pull request by quitterie-lcs**
## Purpose

LDP backends is available for tests.But until now, no service was available in the project.

## Proposal

- [x] add docker-compose service
- [x] add Makefile rule
- [x] add backend
- [x] add push functionality in backend
- [x] add fetch functionality in backend
- [x] add tests for graylog backend
- [x] add logging backend documentation

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -123,6 +124,12 @@ run-es: ## start elasticsearch backend
 	@$(COMPOSE_RUN) dockerize -wait tcp://elasticsearch:9200 -timeout 60s
 .PHONY: run-es
 
+run-graylog: ## start graylog backend
+	@$(COMPOSE) up -d graylog graylog_elasticsearch graylog_mongo
```
Comment (jmaupetit) [inline Makefile:128]: Compose dependency management will do it for you.

```suggestion
	@$(COMPOSE) up -d graylog
```

File: docker-compose.yml [hunk 2]
```diff
@@ -37,6 +37,47 @@ services:
...
+        soft: -1
+        hard: -1
+    ports:
+      - 9300:9300
+    deploy:
+      resources:
+        limits:
+          memory: 1g
+    networks:
+      - default
```
Comment (jmaupetit) [inline docker-compose.yml:65]: Can't we use the existing elasticsearch service to save local resources?

Comment (quitterie-lcs) [reply docker-compose.yml:65]: If we test a transfer from ldp to es, it will fail isn't it?

Comment (jmaupetit) [reply docker-compose.yml:65]: It has to be tested, but I don't think so: ES is a backend used by graylog but graylog will be targeted by our requests.

File: .env.dist [hunk 3]
```diff
@@ -4,6 +4,11 @@ RALPH_APP_DIR=/app/.ralph
 # Uncomment lines (by removing # characters at the beginning of target lines)
 # to define environment variables associated to the backend(s) you need.
 
+# Graylog storage backend
+
+# RALPH_GRAYLOG_HOST=localhost
```
Comment (jmaupetit) [inline .env.dist:9]:
```suggestion
# RALPH_GRAYLOG_HOST=graylog
```

File: tests/backends/logging/test_graylog.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        endpoint.
+        """
+        # pylint: disable=unused-argument
+
+        return """{"bc1c7764-5c7c-4cc0-92b9-ec2759ac1fa0": {"text": "foo"}}"""
+
+    monkeypatch.setattr(GraylogAPI, "get", mock_get)
+    result = api.get_node_id()
+
+    assert uuid.UUID(result)
```
Comment (jmaupetit) [inline tests/backends/logging/test_graylog.py:90]:
```suggestion
    assert uuid.UUID(result)
    assert result == "bc1c7764-5c7c-4cc0-92b9-ec2759ac1fa0"
```

File: tests/backends/logging/test_graylog.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        return "619befa63f959f44ab2ce10a"
+
+    monkeypatch.setattr(
+        "sys.stdin", StringIO("\n".join([json.dumps({"id": idx}) for idx in range(10)]))
+    )
+    monkeypatch.setattr(logging.api, "get_node_id", mock_get_node_id)
+    monkeypatch.setattr(logging.api, "list_inputs", mock_list_inputs)
+    monkeypatch.setattr(logging.api, "activate_input", mock_activate_input)
+
+    assert logging.api.activate_input()
```
Comment (jmaupetit) [inline tests/backends/logging/test_graylog.py:153]:
```suggestion
    assert logging.api.activate_input() == '{"id": "619befa63f959f44ab2ce10a"}'
```

**Review by jmaupetit — COMMENTED** [review]
Looks great :muscle: 

Remarks:

1. Instructions to run/access the graylog server are missing in the project's README
2. Should we pre-configure the graylog server in development (create and activate inputs, create streams)?

File: tests/backends/logging/test_graylog.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    "headers",
+    [
+        {
+            "X-Requested-By": "Learning Analytics Playground",
+            "Content-Type": "application/json",
+        }
+    ],
+)
+def test_graylog_api_get_node_id_method(monkeypatch, url, username, password, headers):
+    """Tests"""
```
Comment (SergioSim) [inline tests/backends/logging/test_graylog.py:75]:
```suggestion
    """Tests that the `get_node_id` method returns the expected node UUID.""
```

File: tests/backends/logging/test_graylog.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+        {
+            "X-Requested-By": "Learning Analytics Playground",
+            "Content-Type": "application/json",
+        }
+    ],
+)
+def test_graylog_api_good_instantiation(url, username, password, headers):
+    """Tests if the `GraylogAPI` is correctly instantiated if a request is
+    successful on the Graylog server.
+    """
```
Comment (SergioSim) [inline tests/backends/logging/test_graylog.py:53]:
```suggestion
    """Tests the GraylogAPI instantiation."""
```

**Review by SergioSim — COMMENTED** [review]
Beautiful work!
Have only minor suggestions.

File: src/ralph/backends/logging/graylog.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        self.port = port
+        self.external_port = external_port
+        self.username = username
+        self.password = password
+
+        self.gelf_logger = logging.getLogger("gelf")
+        self.gelf_logger.setLevel(logging.INFO)
+
+        self.api = GraylogAPI(
+            url=f"http://{self.host}:{self.external_port}",
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/backends/logging/graylog.py:129]: Indeed, I hesitated bey week these ideas and keeping the backend parameters as input as they are also used elsewhere in the project. But after a few times, I found the url a more pythonic solution.

File: src/ralph/backends/logging/graylog.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            username=self.username,
+            password=self.password,
+            headers={
+                "X-Requested-By": "Learning Analytics Playground",
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/backends/logging/graylog.py:133]: Oupssssssss :hand_over_mouth:

File: src/ralph/backends/logging/graylog.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+        with requests.put(
+            f"{self.url}{endpoint}", auth=self._auth, headers=self.headers
+        ) as result:
+            result.raise_for_status()
+
+            return result
+
+    def get_node_id(self):
+        """Returns node id of the Graylog cluster."""
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/backends/logging/graylog.py:74]: No it returns the key of the first listed node, not the dictionary

File: src/ralph/backends/logging/graylog.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+            return result
+
+    def post(self, endpoint, data):
+        """POST method."""
+
+        with requests.post(
+            urllib.parse.urljoin(self.base_url, endpoint),
+            json=data,
+            auth=self._auth,
+            headers=self.headers,
```
Comment (jmaupetit) [inline src/ralph/backends/logging/graylog.py:64]: You should also stream the content and yield the response!

File: src/ralph/backends/logging/graylog.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+            return result.text
+
+    def put(self, endpoint):
+        """PUT method."""
+
+        with requests.put(
+            urllib.parse.urljoin(self.base_url, endpoint),
+            auth=self._auth,
+            headers=self.headers,
```
Comment (jmaupetit) [inline src/ralph/backends/logging/graylog.py:76]: ditto

File: src/ralph/backends/logging/graylog.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+    def activate_input(self, input_id):
+        """Activates a launched input."""
+
+        return self.put(f"/api/system/inputstates/{input_id}")
+
+    def search_logs(self, params):
+        """Returns logs matching given `params` parameters."""
+
+        return self.get("/api/search/universal/relative", params=params)
```
Comment (jmaupetit) [inline src/ralph/backends/logging/graylog.py:110]: I did not found documentation for this endpoint. Can you point me to it?

**Comment by quitterie-lcs** [comment]
[Ajout backend Graylog dans ralph](https://trello.com/c/Mm6sR5RS/7-ajout-backend-graylog-dans-ralph)

**Pull request closed without merging**